src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs

```diff
@@ -14,22 +14,37 @@ public static int CountDigits(UInt128 value)
         {
             ulong upper = value.Upper;
 
-            if (upper < 5)
+            // 1e19 is    8AC7_2304_89E8_0000
+            // 1e20 is  5_6BC7_5E2D_6310_0000
+            // 1e21 is 36_35C9_ADC5_DEA0_0000
+
+            if (upper == 0)
             {
+                // We have less than 64-bits, so just return the lower count
                 return CountDigits(value.Lower);
             }
 
-            int digits = 19;
+            // We have more than 1e19, so we have at least 20 digits
+            int digits = 20;
 
             if (upper > 5)
             {
-                digits++;
+                // ((2^128) - 1) / 1e20 < 34_02_823_669_209_384_635 which
+                // is 18.5318 digits, meaning the result definitely fits
+                // into 64-bits and we only need to add the lower digit count
+
```

Comment:
```suggestion
                // Here we divide the value by 1e20 and count the remaining digits from the result, 
                // all of which should fit within value.Lower, as shown:
                // ((2^128) - 1) / 1e20 is less than 34_02_823_669_209_384_635, 
                // which has 18.5318 digits, meaning the result of the division 
                //definitely fits into 64-bits for any value.

```
Adjusted for clarity, feel free to tweak further.

Reply:
I'll submit a separate PR to cleanup the comment more.